### ISSUE TITLE
Feature Request - Audit the use of Enter-PSHostProcess

### ISSUE THREAD

**Issue by sassdawe** [PowerShell/PowerShell#14634]
## Summary of the new feature/enhancement

As a security professional I want to audit the use of Enter-PSHostProcess on my network to gain better visibility into the abuse of different services running PowerShell by an insider or an external attacker. And I would like to audit this without turning on script-block logging which would generate a lot more noise than a more targeted auditing.

## Proposed technical implementation details (optional)

A dedicated <log> (please insert platform appropriate term) generated every time when Enter-PSHostProcess is used, with metadata containing the source account + process, and the target account + process + command line (it could be a scheduled task too)

**Comment by doctordns** [comment]
I like the basic idea. You might also want to audit ``Invoke-Command -Session`` as well. To cover the case where a user runs something in a session vs entering it. Perhaps different log events for an interactive session and a remote session.

**Comment by iSazonov** [comment]
I see already the request for two cmdlets. Someone might want other cmdlets. As a result, we will come to the conclusion that we need to log everything.

**Comment by iSazonov** [comment]
@sassdawe I understand your request but I have strong doubts that it should be in PowerShell. All impersonating events will be logged in Windows (if we say about Windows) Security log and any security monitoring system will raise alert if this event is illegal. Also, this monitoring system can find related events in other logs including PowerShell.

**Comment by doctordns** [comment]
> I see already the request for two cmdlets. Someone might want other cmdlets. As a result, we will come to the conclusion that we need to log everything.

Is that a bad thing?

As long as it's configurable is that really a problem?

**Comment by sassdawe** [comment]
@iSazonov, you're right! Of course. I should have search this first! 

EventID 4624 captures when Enter-PSHostProcess is used. But it doesn't tell which process is connecting to which, just the two user accounts, and svchost.exe. 

Some information is captured, but not all. Would filling this gap be PowerShell's job on the engine level? It "sounds" easier then asking the OS to provide this information.

**Comment by iSazonov** [comment]
> Some information is captured, but not all

I'd expect Windows logs contain all that you need. The event has an correlation information. You could discuss on Wondows security forums how find correlated events.

**Comment by iSazonov** [comment]
> > I see already the request for two cmdlets. Someone might want other cmdlets. As a result, we will come to the conclusion that we need to log everything.
> 
> Is that a bad thing?
> 
> As long as it's configurable is that really a problem?

I believe PowerShell logs already all needed information. Users only need to enable logging and handle the logs appropriately (which may not be trivial).

**Comment by doctordns** [comment]
Yes - logging exists. But it slows down things. What i9s being asked for is, hopefully, a lot less non-performant.

**Comment by iSazonov** [comment]
> But it slows down things.

If you see real problem with a performance please open new issue with demo and measurements so that we could investigate and fix.

**Comment by sassdawe** [comment]
I believe what @doctordns meant by "but it slows things down" is that correlating different events using some sort of (security) tool makes detection slower or require more actions to be taken for local detection.

I think these are the options:

1. using some sort of SIEM solution create a complex query, which might requires to enable additional logging (this additional logging could have performance impact on the systems being monitored, plus might increase the cost of the SIEM solution)
2. locally, using some advanced queries, which on their own could impact performance depending on how easy is to correlate the events, and if additional logging is requires or not.
3. PowerShell getting an update which would generate one or more predefined and well documented events which could be used on their own, without requiring complex queries to be run. 

Of course, I could be missing something, am I missing something here? Please tell me if I am!

**Comment by doctordns** [comment]
What I also meant is that turning on script block logging has been a perf drain in the past.

**Comment by iSazonov** [comment]
In general. Enter-PSHostProcess can be executed on system non-controlled by you. This convinces me once again that there is no need to add audit to Enter-PSHostProcess. Only target system should log connections.

> Of course, I could be missing something, am I missing something here? Please tell me if I am!

Windows allows you also:
1. Assign a script on events in Event logs.
2. Centralize collecting events on one host. Then you can process its locally or write to a database.
